### PR TITLE
[ncl] Fix most of the TS errors adding types

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

NCL is full of errors.

# How

Went through all of them and fixed them as it made sense in my opinion.

# Test Plan

TS no longer complains that much.

### CODE REVIEW

File: apps/native-component-list/src/navigation/MainNavigators.tsx [hunk 1]
```diff
@@ -29,4 +33,13 @@ export default {
     },
     navigator: ExpoComponentsStackNavigator,
   },
-};
+} as Record<
+  string,
+  | undefined
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/MainNavigators.tsx:38]: Why this `undefined` here? 🤔

File: apps/native-component-list/src/navigation/MainNavigators.tsx [hunk 2]
```diff
@@ -29,4 +33,13 @@ export default {
     },
     navigator: ExpoComponentsStackNavigator,
   },
-};
+} as Record<
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/MainNavigators.tsx:36]: More `regular` approach would be writing:
```ts
const MainNavigators: Record<string, { ...NavigatorType }> = {
  apis: ...,
  components: ...,
}

export default MainNavigators;
```

File: apps/native-component-list/src/navigation/MainNavigators.tsx [hunk 3]
```diff
@@ -29,4 +33,13 @@ export default {
...
+  string,
+  | undefined
+  | {
+      linking: any;
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/MainNavigators.tsx:40]: Shouldn't it be a known type? I mean, isn't `react-navigation` exposing some useful typings?`

File: apps/native-component-list/src/navigation/MainTabNavigator.tsx [hunk 4]
```diff
@@ -52,21 +57,28 @@ export default function MainTabbedNavigator(props: any) {
...
+              <Tab.Screen
+                name={name}
+                key={name}
+                component={screen.navigator}
+                options={screen.navigator.navigationOptions}
+              />
+            );
+          }
+          return null;
+        })}
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/MainTabNavigator.tsx:81]: Well, it is not a nicely written code 😞 
```suggestion
        {Object.entries(Screens)
           .filter(([name, Screen]) => Boolean(Screen))
           .map(([name, Screen]) => (
              <Tab.Screen
                name={name}
                key={name}
                component={screen.navigator}
                options={screen.navigator.navigationOptions}
              />
           ))
        }
```
Why some `Screens` not defined?
I see these are imported from `./MainNavigators` and the object exported there is of some explicitly typed structure. I didn't see any `undefined`s as values there 🤔

File: apps/native-component-list/src/navigation/MainTabNavigator.tsx [hunk 5]
```diff
@@ -77,14 +89,20 @@ export default function MainTabbedNavigator(props: any) {
...
+            <Drawer.Screen
+              name={name}
+              key={name}
+              component={screen.navigator}
+              options={screen.navigator.navigationOptions}
+            />
+          );
+        }
+        return null;
+      })}
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/MainTabNavigator.tsx:105]:
```suggestion
      {Object.entries(Screens)
        .filter(([]) => )
        .map(([name, { navigator }] => (
          <Drawer.Screen
            name={name}
            key={name}
            component={navigator}
            options={navigator.navigationOptions}
          />
        ))
      }
```

File: apps/native-component-list/src/navigation/RootNavigation.tsx [hunk 6]
```diff
@@ -18,8 +18,8 @@ export const linking = {
       main: {
         initialRouteName: 'apis',
         screens: {
-          apis: MainNavigators.apis.linking,
-          components: MainNavigators.components.linking,
+          apis: MainNavigators.apis?.linking,
+          components: MainNavigators.components?.linking,
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/RootNavigation.tsx:22]: Once `undefined` is removed from `MainNavigators` we're ready to revert this change
```suggestion
          apis: MainNavigators.apis.linking,
          components: MainNavigators.components.linking,
```

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx [hunk 7]
```diff
@@ -127,7 +127,7 @@ function ResizeModeSegmentedControl({
 }: {
   onValueChange: (value: ResizeMode) => void;
 }) {
-  const resizeMap = {
+  const resizeMap: Record<string, ResizeMode | undefined> = {
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/AV/VideoPlayer.tsx:130]: Why are you typing sth that is not representing an actual JS state? 🤔 
```suggestion
  const resizeMap: Record<string, ResizeMode> = {
```

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx [hunk 8]
```diff
@@ -151,7 +151,12 @@ function ResizeModeSegmentedControl({
         onChange={event => {
           setIndex(event.nativeEvent.selectedSegmentIndex);
         }}
-        onValueChange={value => onValueChange(resizeMap[value])}
+        onValueChange={value => {
+          const mappedValue = resizeMap[value];
+          if (mappedValue) {
+            onValueChange(mappedValue);
+          }
+        }}
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/AV/VideoPlayer.tsx:159]: If TS is complaining here then you need to adjust it somewhere else. From what I can see here, `value` can only one of known strings (namely, keys from `resizeMap`)
```suggestion
        onValueChange={value => onValueChange(mappedValue)}
```

File: apps/native-component-list/src/screens/FacebookLoginScreen.tsx [hunk 9]
```diff
@@ -57,14 +57,12 @@ export default class FacebookLoginScreen extends React.Component {
         permissions: perms,
       });
 
-      const { type, token } = result;
-
-      if (type === 'success') {
+      if (result.type === 'success') {
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/FacebookLoginScreen.tsx:60]: Why have you changed these lines? 🤔

File: apps/native-component-list/src/screens/HapticsScreen.tsx [hunk 10]
```diff
@@ -1,12 +1,19 @@
...
-import { SectionList, StyleSheet, Text, View, ViewStyle } from 'react-native';
+import { SectionList, SectionListData, StyleSheet, Text, View, ViewStyle } from 'react-native';
 
 import Button from '../components/Button';
 import MonoText from '../components/MonoText';
 import Colors from '../constants/Colors';
 
-const sections = [
+type SectionData =
+  | object
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/HapticsScreen.tsx:10]: That's incorrect. The actual type we need here is an `empty object type`, namely `{}`. I know we have a lint rule preventing this, but in this case it should be that one 😉
Why? Because of https://mariusschulz.com/articles/the-object-type-in-typescript

**Review by bbarthec — COMMENTED** [review]
Some comments left 👍 
In general, thank you for taking a TS round through `ncl` code 💪

File: apps/native-component-list/src/navigation/MainNavigators.tsx (hunk 1 above)
Comment (sjchmiela) [reply apps/native-component-list/src/navigation/MainNavigators.tsx:38]: Because not for every `string` there will be a value? I wanted to make it clear this way.

File: apps/native-component-list/src/navigation/MainNavigators.tsx (hunk 3 above)
Comment (sjchmiela) [reply apps/native-component-list/src/navigation/MainNavigators.tsx:40]: Good call, it's `PathConfig`.

File: apps/native-component-list/src/navigation/MainTabNavigator.tsx (hunk 4 above)
Comment (sjchmiela) [reply apps/native-component-list/src/navigation/MainTabNavigator.tsx:81]: I wanted to make as small changes as possible (keeping `keys.map` as it was before), but then TS complained about the fact that `Screens[name]` could be `undefined` (as per `Record<string, undefined | …>`), so I did the `if ()` so it knows it's not `undefined`.

Your proposed solution doesn't give TS confidence that `screen` is not `undefined`, there's an error.

Maybe the problem lies in the `undefined` that it should not be there (logically I thought it should be there but I may be wrong). 🤷‍♂️

File: apps/native-component-list/src/navigation/MainTabNavigator.tsx (hunk 5 above)
Comment (sjchmiela) [reply apps/native-component-list/src/navigation/MainTabNavigator.tsx:105]: Same comment as https://github.com/expo/expo/pull/10229#discussion_r490968366

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx (hunk 7 above)
Comment (sjchmiela) [reply apps/native-component-list/src/screens/AV/VideoPlayer.tsx:130]: Because not for every `string` there's a value. To remove `undefined` I feel I should do `Record<'stretch' | 'contain' | 'cover', ResizeMode>` (as I said in some other command I may be abusing `Record`).

File: apps/native-component-list/src/screens/HapticsScreen.tsx (hunk 10 above)
Comment (sjchmiela) [reply apps/native-component-list/src/screens/HapticsScreen.tsx:10]: I thought so too, but figured lint rule may know better. Will fix.

File: apps/native-component-list/src/navigation/MainNavigators.tsx (hunk 1 above)
Comment (sjchmiela) [reply apps/native-component-list/src/navigation/MainNavigators.tsx:38]: FWIW [`Record` docs](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeystype) define `Keys` as an explicit union type.

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx (hunk 7 above)
Comment (bbarthec) [reply apps/native-component-list/src/screens/AV/VideoPlayer.tsx:130]: Why do we need to specify the type for this variable like this in the first place? 🤔
Why not to go with implicit type inferred from the object decalration?
```ts
const resizeMap = {
  ...
};
```

Comment (bbarthec) [reply apps/native-component-list/src/screens/AV/VideoPlayer.tsx:130]: You can be more specific and do it like this:
```ts
const resizeMap = {
  ...
} as const
```

Comment (sjchmiela) [reply apps/native-component-list/src/screens/AV/VideoPlayer.tsx:130]: Because in `onValueChange` callback the _new value_ is of type `string` for which there may be no mapping in the `resizeMap` map according to TS (it knows `resizeMap` only defines values for three specific literals):
Zrzut ekranu 2020-09-18 o 16 10 41

Comment (bbarthec) [reply apps/native-component-list/src/screens/AV/VideoPlayer.tsx]: And this is perfectly logical, because `SegmentedControl` is given `values={data}` and `data = Object.keys(...) // string[]`. If `SegmentedControl` is properly inferring the type from the available values then it must be producing `string` type in `onChange` callback. I wonder if changing `values={data as (typeof keyof Screens)[]}` is not going to fix the issue with later casting 😉

File: apps/native-component-list/src/navigation/MainTabNavigator.tsx [hunk 11]
```diff
@@ -52,14 +57,15 @@ export default function MainTabbedNavigator(props: any) {
...
+        // bottom-tabs props
         tabBarOptions={{
           style: {
             backgroundColor: Colors.tabBar,
           },
           activeTintColor: Colors.tabIconSelected,
           inactiveTintColor: Colors.tabIconDefault,
         }}>
-        {Object.keys(Screens).map(name => (
+        {(Object.keys(Screens) as (keyof typeof Screens)[]).map(name => (
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/MainTabNavigator.tsx:68]: I still believe that using
```suggestion
        {Object.entries(Screens).map(([name, Component]) => (
```
would solve TS errors and wouldn't force us to use this nasty casting 😉

File: apps/native-component-list/src/navigation/MainTabNavigator.tsx [hunk 12]
```diff
@@ -77,8 +83,8 @@ export default function MainTabbedNavigator(props: any) {
       drawerContent={props => <CustomDrawerContent {...props} hideLabels={isTablet} />}
       drawerStyle={{ width: isLargeScreen ? undefined : 64 + left }}
       drawerType="permanent">
-      {Object.keys(Screens).map(name => (
-        <Drawer.Screen
+      {(Object.keys(Screens) as (keyof typeof Screens)[]).map(name => (
```
Comment (bbarthec) [inline apps/native-component-list/src/navigation/MainTabNavigator.tsx:86]: Same here 😉 
I know that you don't like to change both syntax/semantics (js ➡️ ts) as well actual logic in one PR, but at this point I believe it's irrelevant to keep this distinction, because of:
- you can write exactly the same logic using other language constructs
- fixing TS errors sometimes even suggests these better tailored semantics for some cases (like this one 😉)

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx [hunk 13]
```diff
@@ -144,14 +143,14 @@ function ResizeModeSegmentedControl({
         flex: 1,
       }}>
       <SegmentedControl
-        values={data}
+        values={Object.keys(resizeMap)}
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/AV/VideoPlayer.tsx:146]:
```suggestion
        values={Object.keys(resizeMap) as (keyof typeof resizeMap)[]}
```

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx [hunk 14]
```diff
@@ -144,14 +143,14 @@ function ResizeModeSegmentedControl({
...
         fontStyle={{ color: Colors.tintColor }}
         selectedIndex={index}
         tintColor={'white'}
         onChange={event => {
           setIndex(event.nativeEvent.selectedSegmentIndex);
         }}
-        onValueChange={value => onValueChange(resizeMap[value])}
+        onValueChange={value => onValueChange(resizeMap[value as keyof typeof resizeMap])}
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/AV/VideoPlayer.tsx:153]:
```suggestion
        onValueChange={value => onValueChange(resizeMap[value])}
```
Given the change from the other suggestion regarding this Component, it should work like this. _Assuming `SegmentedControl` is well typed_

**Review by bbarthec — APPROVED** [review]
Nice 🎉 
Keep up the hard work 👍 
Left further comments on this `Object.keys` matter, but overall I'm approving it 👍

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx (hunk 7 above)
Comment (sjchmiela) [reply apps/native-component-list/src/screens/AV/VideoPlayer.tsx]: I know it is perfectly logical, I just post the error here for you to see exactly what I see. `SegmentedControl` does not infer types (if it would I would have used it). `values={data as (typeof keyof Screens)[]}` wouldn't change anything (we would still need to cast in `onChange`).

File: apps/native-component-list/src/navigation/MainTabNavigator.tsx (hunk 12 above)
Comment (sjchmiela) [reply apps/native-component-list/src/navigation/MainTabNavigator.tsx:86]: I didn't notice `Object.entries` would solve this problem (my bad, I tried it when `Screens` was still a `Record<…, undefined | …>`, which obviously didn't help at all then), I agree that if a slightly different code produces the same result, is as readable as its previous version and solves a types problem it's worth changing it. 👍

File: apps/native-component-list/src/screens/AV/VideoPlayer.tsx (hunk 7 above)
Comment (sjchmiela) [reply apps/native-component-list/src/screens/AV/VideoPlayer.tsx]: Adding it would make sense if we wanted to have like, two places where we define that values in `values` and in `onChange` are expected and required to be `typeof keyof resizeMap`, but it is also (imho) error-prone (there are two places instead of one and we can't really guarantee on type level without `SegmentedControl` generics that casting in `onChange` will not be an error.

All in all I love types too, but I wouldn't say that omitting one `if (resizeMap[value])` is worth the inconveniences that trying to make this safe on type level which is currently impossible/very hard to do properly (given this 3rd party library type definitions).
